ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateStateGenesis.java

```diff
@@ -0,0 +1,101 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.privacy;
+
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.EvmAccount;
+import org.hyperledger.besu.ethereum.core.MutableAccount;
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+import org.hyperledger.besu.ethereum.core.Wei;
+import org.hyperledger.besu.ethereum.core.WorldUpdater;
+import org.hyperledger.besu.ethereum.privacy.group.OnChainGroupManagement;
+import org.hyperledger.besu.plugin.data.PrivacyGenesis;
+import org.hyperledger.besu.plugin.services.privacy.PrivacyGroupGenesisProvider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.units.bigints.UInt256;
+
+public class PrivateStateGenesis {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Boolean isOnchainPrivacyEnabled;
+  private final PrivacyGroupGenesisProvider privacyGroupGenesisProvider;
+
+  public PrivateStateGenesis(
+      final Boolean isOnchainPrivacyEnabled,
+      final PrivacyGroupGenesisProvider privacyGroupGenesisProvider) {
+    this.isOnchainPrivacyEnabled = isOnchainPrivacyEnabled;
+    this.privacyGroupGenesisProvider = privacyGroupGenesisProvider;
+  }
+
+  public void applyGenesisToPrivateWorldState(
+      final MutableWorldState disposablePrivateState,
+      final WorldUpdater privateWorldStateUpdater,
+      final Bytes privacyGroupId,
+      final long blockNumber) {
+
+    final PrivacyGenesis genesis =
+        privacyGroupGenesisProvider.getPrivacyGenesis(privacyGroupId, blockNumber);
+
+    if (genesis.getAccounts().size() > 0) {
+      LOG.info("Applying {} privacy accounts onto private genesis", genesis.getAccounts().size());
+
+      genesis
+          .getAccounts()
+          .forEach(
+              (genesisAccount) -> {
+                final MutableAccount account =
+                    privateWorldStateUpdater
+                        .createAccount(Address.fromPlugin(genesisAccount.getAddress()))
+                        .getMutable();
+
+                account.setNonce(genesisAccount.getNonce());
+                account.setBalance(Wei.fromQuantity(genesisAccount.getBalance()));
+                account.setCode(genesisAccount.getCode());
+                account.setVersion(genesisAccount.getVersion());
+
+                genesisAccount.getStorage().forEach(account::setStorageValue);
+              });
+    }
+
+    if (isOnchainPrivacyEnabled) {
+      // inject management
+      final MutableAccount managementPrecompile =
```

Comment:
Well, technically I think it's a precompile because it's compiled and deployed to world state in advance! But I take your point, it's confusing to mix the naming with the other precompiles in besu. renamed.